c2rust-analyze/src/main.rs

```diff
@@ -449,13 +449,117 @@ fn run(tcx: TyCtxt) {
+                {
+                    let func_ty = func.ty(&mir.local_decls, tcx);
+
+                    // Finds casts associated with the special functions malloc, calloc,
+                    // realloc, and free. For the first three, there is expected to be
+                    // a cast from void* to an arbitrary type in the subsequent block.
+                    // For realloc and free, there is expected to be a cast to void*
+                    // from an arbitrary type in the same block.
+                    let mut handle_c_void_cast = |void_ptr: &Place<'tcx>,
+                                                  find_cast: &dyn Fn() -> Option<(
+                        Place<'tcx>,
+                        Place<'tcx>,
+                    )>| {
+                        let deref_ty = void_ptr
+                            .ty(acx.local_decls, acx.tcx())
+                            .ty
+                            .builtin_deref(true)
+                            .unwrap()
+                            .ty
+                            .kind();
+                        assert_matches!(deref_ty, TyKind::Adt(adt, _) => {
+                            assert_eq!(tcx.def_path(adt.did()).data[0].to_string(), "ffi");
+                            assert_eq!(tcx.item_name(adt.did()).as_str(), "c_void");
+                        });
+
+                        if let Some((casted, castee)) = find_cast() {
+                            if castee == *void_ptr {
+                                // This is a special case for types being casted from *c_void to a pointer
+                                // to some other type, e.g. `let foo = malloc(..) as *mut Foo;`
+                                // carry over the pointer id of *c_void, but match the pointer ids
+                                // of the casted-to-type for the rest
+                                acx.c_void_casts.0.insert(castee, casted);
+                            }
+                        }
+                    };
+
+                    let get_cast_op = |s: &Statement<'tcx>| match &s.kind {
+                        StatementKind::Assign(x) => match &x.1 {
+                            Rvalue::Cast(_, op, _) => Some((
+                                x.0,
+                                op.place().expect(
+                                    "Rvalue::Cast is not Operant::Constant so it has a Place",
+                                ),
+                            )),
+                            _ => None,
+                        },
+                        _ => None,
+                    };
+
+                    let find_last_cast_curr_block = || {
+                        bb_data
+                            .statements
+                            .iter()
+                            .rev()
+                            .find_map(|s| get_cast_op(s))
+                            .map(|(casted_to, casted_from)| (casted_from, casted_to))
+                    };
```

Comment:
We need a check somewhere to ensure that the LHS of the cast isn't modified between the cast and the `free` call.  With the current implementation, the correctness of our `free` handling depends on the fact that rustc currently produces suboptimal MIR for code like this:
```Rust
let mut q = p as *mut c_void;
q = x;
free(q);
```
The corresponding MIR:
```Rust
_4 = p;
q = move _4 as *mut libc::c_void (PtrToPtr);
_5 = x;
q = move _5;
_7 = q;
_6 = free(move _7) -> bb1;
```
If rustc changed to produce `_6 = free(q) -> bb1;` (eliminating the unnecessary temporary `_7`), then I believe our current implementation would wrongly recognize this as `free(_4)`.

Checking whether a statement could modify a particular `Place` is nontrivial (particularly if the `Place` involves `Deref` projections), but currently rustc always uses a temporary for the LHS of the cast in the desired `free(ptr as *mut c_void)` pattern, so I think we can conservatively treat the `Place` as unmodified only if (1) it's just a `Local`, with no projections, and (2) that local isn't mentioned in the LHS or RHS of any statements between the cast and the call.

Reply:
I've addressed this in #848